### DISCUSSION TITLE
sensor crop, frame resolution, packing

### DISCUSSION THREAD

**Discussion by ballen4705** [raspberrypi/picamera2#585] [Q&A]
I have a question about the relationship between sensor crop, frame resolution, and packing/alignment.

My camera is a RP HQ, which offers four modes.   The mode I want is `picam2.sensor_modes[2]`, for which the relevant entries are:
```
'crop_limits': (0, 0, 4056, 3040)
'size': (2028, 1520)
```
The relevant part of my config is:
```
raw  = {"size":(2028,1520)}
main = {"size":(2028,1520), "format":"YUV420"}
```
Although I do not use raw directly, I specify the raw size to ensure that main contains the full uncropped 4056x3040 field of view (FOV). 

The motivation for my question is that I'd like to have an efficient packing/alignment for the data. Appendix A of the "Picamera2 Library" says that YUV420 is best aligned on 64-pixel boundaries. Indeed, `picam2.align_configuration(config)` suggests 1984 x 1520 pixels for main, corresponding to
```
raw  = {"size":(2028,1520)}
main = {"size":(1984,1520), "format":"YUV420"}
```
**My issue**: this has the wrong aspect ratio; about 1984 x 1487 pixels would be correct.

**My question**: how does libcamera/picamera2 cope with this?  Does it:
- crop the image, dropping 22 columns of pixels on the far left side of the sensor and 22 columns on the far right side?
- interpolate/scale to produce a 1984 x 1520 pixel image where the full FOV is mapped onto that rectangle, but the aspect ratio is no longer correct?
- fit the full field of view into a 1984x1487 subset of the 1984x1520 output, putting zeros or 255 or random numbers into the final 33 rows?
- do something else?

To pose the question differently,  suppose that I set
```
raw  = {"size":(2028,1520)}
main = {"size":(2048,1536)}
```
which aligns main on 64-pixel boundaries and is very close to the sensor's aspect ratio (almost exactly 2048x1535). Is the full FOV mapped onto the 2048x1536 array?  Or is the final row of the image set to nonsense values?

I'm not sure if this matters for photography, but it is important for image analysis.

Cheers,
   Bruce

**Comment by davidplowman** [comment]
To answer these kinds of questions, it might be helpful, after configuring Picamera2 (with `picam2.configure(config)`) to check `picam2.camera_controls['ScalerCrop'][2]`.  This will give you the rectangle it has  chosen to crop from the sensor array.

For the example above, where the main stream is 1984x1520 in size, `picam2.camera_controls['ScalerCrop][2]` reports `(44, 0, 3968, 3040)`. So it's cropping out the largest part of the sensor that it can to provide the aspect ratio that was requested (without "squishing" anything in only one dimension).

For YUV420 output you can't have an image that is 1487 pixels high, but if you ask for a 1984x1488 output the scaler crop reports `(2, 0, 4052, 3040)`, so you have, to all intents and purposes, the full FoV.

Generally there shouldn't be too much reason to align configurations unless you have downstream software that can't deal with arrays that are not fully contiguous (i.e. may have padding at the ends of rows), though others may have more practical experience of this.

**Comment by RainerHeintzmann** [comment]
I have a related issue: If I use the `ScalerCrop` properties, they work fine and crop the video stream. However when I acquire an image via `picam2.capture_array("raw")` this seems to ignore all the cropping and just returns the full array.
However I was hoping the cropping to a small region of interest would allow me to stream (and save) raw data faster than working with the large images and cropping them with python code.
I would like to use the camera for speckle imaging in astro-photography, so speed is important and it needs to be uncompressed data (or lossless compression).

**Comment by davidplowman** [comment]
The `ScalerCrop` is purely a digital process. It takes a potion of the image that has already been delivered by the sensor, and then only sends that further down the image processing pipeline. As a consequence, there is no change to the raw data read from the sensor (you still get everything), and you can't use this to speed up the framerate.

Most sensors, and certainly all of our officially supported ones, only allow a fixed selection of sensor modes, which in turn determines the raw buffer size and characteristics. `ScalerCrop` is typically used to implement digial pan/zoom within that image.

Having said all that, there may be ways to help you.

1. What kind of (raw buffer) resolution/framerate are you hoping for? Depending on your platform and storage medium, fairly high pixel throughputs are achievable (e.g. [here](https://forums.raspberrypi.com/viewtopic.php?t=397255)). 

2. Although we can't boost the sensor framerate, if saving excessively large DNG files is a  bottleneck, it would be possible to use ("trick") the `Picamera2.helpers.save_dng` method into saving cropped DNG files.

**Comment by RainerHeintzmann** [comment]
My coworker was using some old style code (https://github.com/Hermann-SW/fork-raspiraw) and achieved about 1000 fps for small ROIs as far as I remember.
I was hoping that the same is also possible with the new picam2 API.
Yes, trying the `save_dng()` function is a good idea. 

Independent on where the processing occurs, I still don't quite understand why `ScalerCrop` seemed to work in the video mode but not in the raw single frame acquisition code (`picam2.capture_array("raw")`), where it always returned the whole image.
Maybe `picam2.capture_array("raw")` should be replaced with a different command?

I assume that ScalerCrop is a pure CPU-based software (C++?) on the computer. In this case I cannot hope for much higher speeds using or not using it.
But if this was implemented on the firmware and cops before sending the data over the cable, it would massively reduce the data transfer and therefore be really promising for high-speed data capture.

**Reply by davidplowman** [reply]
> My coworker was using some old style code (https://github.com/Hermann-SW/fork-raspiraw) and achieved about 1000 fps for small ROIs as far as I remember. I was hoping that the same is also possible with the new picam2 API. Yes, trying the `save_dng()` function is a good idea.

That code works by changing the sensor driver to provide much smaller readout resolutions from the sensor, but the essence is still that you can't speed up the framerate without providing new, faster sensor modes that read out fewer pixels.

I thought there is a version of something like that for the libcamera world, though I can't remember where. I think it's a similar piece of "hackery" (I use the word as a compliment!). If you asked on the camera forum someone there might be able to point you in the right direction.

> 
> Independent on where the processing occurs, I still don't quite understand why `ScalerCrop` seemed to work in the video mode but not in the raw single frame acquisition code (`picam2.capture_array("raw")`), where it always returned the whole image. Maybe `picam2.capture_array("raw")` should be replaced with a different command?

I don't really understand the question here, I'm afraid. `ScalerCrop` always behaves the same. It has no effect on the readout from the sensor, and can only select from what the sensor is already delivering.

If you're observing that you sometimes get different readout sizes from the sensor depending on what output image resolution you request, then this is true. The sensor mode will be selected automatically for you (and frequently not be what you expect) unless you explicitly pass the parameters for the mode that you do want. As I'm sure you know, you can check what sensor modes are available with `rpicam-hello --list-cameras`.

> 
> I assume that ScalerCrop is a pure CPU-based software (C++?) on the computer. In this case I cannot hope for much higher speeds using or not using it. But if this was implemented on the firmware and cops before sending the data over the cable, it would massively reduce the data transfer and therefore be really promising for high-speed data capture.

`ScalerCrop` is implemented by the hardware image signal processor on main processor. For it to happen early enough to allow higher framerates, then there's only the sensor that can do it, which means, effectively, creating new sensor modes and implementing them in the driver.

If you can say more about what you're doing - the model of Pi, the sensor, the framerates/resolution that you want vs. what you are actually achieving, then it might be possible to make some more precise suggestions.
